File: 112-path-sum/112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    bool Traverse(TreeNode* root, int targetSum, int curr_sum){
        
        if(root==NULL)
            return false;
        
        curr_sum+=root->val;
        
        if(curr_sum==targetSum && root->left==NULL && root->right==NULL)
            return true;
        
        return Traverse(root->left, targetSum, curr_sum) || Traverse(root->right, targetSum, curr_sum);
        
    }
    
    bool hasPathSum(TreeNode* root, int targetSum) {
        
        if(root==NULL)
            return false;
        
        return Traverse(root, targetSum, 0);
        
    }
};

File: 112-path-sum/112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def Traverse(self, root, targetSum, curr_sum):
        
        if(root==None):
            return False
        
        curr_sum+=root.val
        
        if(curr_sum==targetSum and (root.right==None and root.left==None)):
            return True
        
        return self.Traverse(root.right, targetSum, curr_sum) or self.Traverse(root.left, targetSum, curr_sum)
    
    
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        if(root==None):
            return False
        
        return self.Traverse(root, targetSum, 0)
        